### ISSUE TITLE
Investigate and tests FreeIPA 2FA

### ISSUE THREAD

**Issue by sfinn85** [fedora-infra/noggin#17]

**Comment by abompard** [comment]
It would be nice to have more details on that one. What should we investigate and test exactly? How should that be integrated in Securitas?

**Comment by sfinn85** [comment]
@abompard going to break this down further

**Comment by relrod** [comment]
@abompard had some questions around this, so I coded up a POC of what it could look like in the [otp branch](https://github.com/fedora-infra/securitas/tree/otp) ([full diff](https://github.com/fedora-infra/securitas/compare/4eafd63648e10409a42fea222dd65446f32fce90%5E%5E%5E...otp)). It needs work (handle error cases, add tests, make it prettier, etc) but gives an idea of how the API calls work...

A few notes/questions/thoughts, mostly on policy:

* IPA lets users manage their own tokens, but we need to decide how we want enforcement to work. Right now with current FAS, we only require 2fa for sudo in Fedora Infrastructure machines. Do we want to keep this policy, or expand it to all access?
* If we want it for sudo only, perhaps @tiran can give some insight on the best way to approach that on the IPA side.
* If we want it for more, how do we communicate that to people? Perhaps we let people opt in for a while if they want it, and announce that later on we'll make it required? But if we do that, we'll need a way to make it still required for sudo at least. @puiterwijk might have some feedback to provide here.

**Comment by puiterwijk** [comment]
I personally think that we should make it so that if people enroll an OTP, we enforce token+password login (i.e. require OTP for anything), which is a user-level setting in the IPA user object.

**Comment by tiran** [comment]
FreeIPA implements selective 2FA policy with a Kerberos/GSSAPI feature called "auth indicators". AFAIK auth indicators only apply to authentication requests. I don't know if it's also possible to restrict HBAC or sudo with 2FA.

To restrict access to a host, you have to enable OTP and password auth (either globally or per-user), then modify the host with ``ipa host-mod  --auth-ind=otp hostname``.

To require OTP, only enable OTP. This makes OTP mandatory for users with an OTP token, Users without an OTP token enrolled are still able to authenticate with a password.

**Comment by tiran** [comment]
By the way: If OTP is required, then you cannot remove the last token. Removal of the last token is blocked by IPA's plugin for LDAP/KRB5.

**Comment by sfinn85** [comment]
Need to reach out to Rick to see if he will continue working on this or will stop and work on colo move

**Comment by sfinn85** [comment]
May lead to more user stories - daily stand up update 2/24/20

**Comment by nirik** [comment]
Some thoughts:

We have the following things we want people to auth for: 
* sudo (admin only)- I guess we will be using sssd for this? can it use otp?
* ssh login to hosts (admin only) - this might be done with certs and we don't care about otp? or can we use otp here somehow?
* https login to applications(users) - those will be via ipsilon currently, can we require otp here?
* https pushing to git (users/packagers) - this is a pagure token currently, do we want to change that?
* https login to securitas - this will automatically use password + otp if there's one enrolled right?

I'd be ok enabling otp globally, but we may want to remind/nag/note to users that they should we nag people to enroll several tokens, so if they don't have access to one they can still login?
 
How do we want to handle the case of a lost/broken/missing token? I guess via admin ticket? But they can't login if we require otp for https login... so email? but then we need to have some way to authenticate them? Or do we just say 'sorry, make a new account' ?
